pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java

```diff
@@ -235,20 +237,36 @@ private CompletableFuture<Reader<T>> readAllExistingMessages(Reader<T> reader) {
         AtomicLong messagesRead = new AtomicLong();
 
         CompletableFuture<Reader<T>> future = new CompletableFuture<>();
-        readAllExistingMessages(reader, future, startTime, messagesRead);
+        reader.getLastMessageIdsAsync().thenAccept(lastMessageIds -> {
+            Map<String, TopicMessageId> maxMessageIds = new HashMap<>();
+            lastMessageIds.forEach(topicMessageId -> {
+                maxMessageIds.put(topicMessageId.getOwnerTopic(), topicMessageId);
+            });
+            readAllExistingMessages(reader, future, startTime, messagesRead, maxMessageIds);
```

Comment:
It's better to make `maxMessageIds` that passed to the `readAllExistingMessages` immutable to prohibit other threads from getting a chance to update the HashMap.

Reply:
We need to check whether all the partitions have been read completely. So, we need to do an update operation to the HashMap after a partition has been read completely. This operation could be removing this partition from the map or adding a flag to mark that the partition has been read completely.
How about changing it to `ConcurrentHashMap`?

Reply:
The solution you mentioned in https://github.com/apache/pulsar/pull/21417#discussion_r1372545861 could be helpful.